### PR TITLE
Add new Docker based build system

### PR DESCRIPTION
**Pull request by p2004a**
First PR that introduces the new docker based engine build system, following up from previous draft PR https://github.com/beyond-all-reason/spring/pull/1707

As written in previous PR, and in commit, the new system is primarily targeting efficiency:
- Smaller and quicker to build Docker build environment images
- Storage of Docker images in registry instead of cache that expires
- Leverages much more performant namespace.so GitHub actions runners
- Each step is optimized for parallelism to leverage new runners
- Heavy cross branch, persistent, caching of compilation objects

### CODE REVIEW

File: .github/workflows/engine-build.yml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+          EOF
+      - name: Save
+        if: github.event_name != 'pull_request'
+        uses: namespace-actions/upload-artifact@v0
+        with:
+          name: output-${{ matrix.system }}
+          path: ./artifacts
+      - name: Unmount bazel remote overlay
+        if: always() && steps.mount-overlay.outcome == 'success'
```
[the remark this answers is not in the document]
Comment (p2004a) [reply .github/workflows/engine-build.yml:144]: The `always()` will make sure the step runs also when any of the previous steps failed. I agree it's not the best readability but that's how it's done in GitHub actions

File: .github/workflows/engine-build.yml [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        if: github.event_name == 'pull_request'
+        run: |
+          sudo apt-get install --yes fuse-overlayfs
+          sudo tar --acls --xattrs --xattrs-include='*' -xf bazel-remote-data-overlay.tar || mkdir bazel-remote-data-overlay
+          mkdir -p overlay-workdir bazel-remote-data-merged
+          sudo fuse-overlayfs -o lowerdir=bazel-remote-data,upperdir=bazel-remote-data-overlay,workdir=overlay-workdir bazel-remote-data-merged
+      - name: Start remote ccache fetcher
+        uses: JarvusInnovations/background-action@2428e7b970a846423095c79d43f759abf979a635  # v1.0.7
+        env:
+          BAZEL_REMOTE_S3_ENDPOINT: ${{ vars.R2_ACCOUNT_ID }}.r2.cloudflarestorage.com
```
[the remark this answers is not in the document]
Comment (p2004a) [reply .github/workflows/engine-build.yml:82]: 1. Ownership is BAR org and it's paid by BAR org just like runners will be
2. Me + Admins are PoCs, just like for any other infrastructure things
3. It's reliable, R2 is designed to be used for interactive traffic. In addition both bazel-remote and ccache offer configurable bailout when they just switch back to compilation if remote endpoint is not available. Currently it's using defaults afaik ~5s.

File: .github/workflows/engine-build.yml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+              --dir bazel-remote-data${{ github.event_name == 'pull_request' && '-merged' || '' }} \
+              --s3.auth_method access_key \
+              --s3.region auto \
+              --max_size 5 \
+              --num_uploaders ${{ github.event_name == 'pull_request' && '0' || '100' }} \
+              --disable_http_ac_validation \
+              --http_address 127.0.0.1:8085 \
+              --access_log_level none &
+          wait-on: |
+            http-get://127.0.0.1:8085/cas/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```
[the remark this answers is not in the document]
Comment (p2004a) [reply .github/workflows/engine-build.yml:111]: Will add a comment when I get to PC, it's a special "empty object" in bazel-remote that can be exactly used for availability checks.

File: docker-build-v2/build.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+docker run -it --rm \
+    -v /etc/passwd:/etc/passwd:ro \
+    -v /etc/group:/etc/group:ro \
+    --user=$(id -u):$(id -g) \
+    -v $(pwd):/build/src:ro \
+    -v $(pwd)/.cache/ccache-$os:/build/cache:rw \
+    -v $(pwd)/build-$os:/build/out:rw \
+    $image \
+    bash /build/src/docker-build-v2/scripts/build.sh "${@:2}"
```
[the remark this answers is not in the document]
Comment (p2004a) [reply docker-build-v2/build.sh:29]: I'm considering splitting the internal build.sh into configure.sh and compile.sh, but also already using different naming would be fine, I'm open to suggestions.

File: docker-build-v2/amd64-linux/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,53 @@
+FROM docker.io/ubuntu:18.04
```
Comment (sprunk) [inline docker-build-v2/amd64-linux/Dockerfile:1]:
```suggestion
# You can't use the build on an older Linux system due to system libs linkage,
# so pick a relatively old one instead of freshest so that everybody can play.
FROM docker.io/ubuntu:18.04
```

**Comment by lhog** [comment]
I skimmed over the code, looks nice!
Will post a few questions in the comments.

File: .github/workflows/engine-build.yml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  #   paths-ignore:
+  #     - 'doc/**'
+jobs:
+  build-engine:
+    strategy:
+      matrix:
+        system:
+          - amd64-linux
+          - amd64-windows
+    runs-on:
```
Comment (lhog) [inline .github/workflows/engine-build.yml:43]: Can you summarize what these below mean?

File: docker-build-v2/amd64-linux/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ && pip3 install scikit-build \
+ && pip3 install cmake==3.25.*
+
+WORKDIR /build
+RUN mkdir src cache out artifacts && chmod a+rwx cache out artifacts
+
+# Fetch library dependencies and configure resolution
+RUN git clone --depth=1 https://github.com/beyond-all-reason/spring-static-libs.git -b 18.04 spring-static-libs
+ENV PKG_CONFIG_LIBDIR=/build/spring-static-libs/lib/pkgconfig
+ENV PKG_CONFIG="pkg-config --define-prefix --static"
```
Comment (lhog) [inline docker-build-v2/amd64-linux/Dockerfile:41]: Did we ever need `PKG_CONFIG_LIBDIR` or `PKG_CONFIG` before?

File: .github/workflows/engine-build.yml (hunk 6 above)
Comment (p2004a) [reply .github/workflows/engine-build.yml:43]: Added comments

File: docker-build-v2/amd64-linux/Dockerfile (hunk 5 above)
Comment (p2004a) [reply docker-build-v2/amd64-linux/Dockerfile:4]: Adjusted text slightly and commited.

File: docker-build-v2/amd64-linux/Dockerfile (hunk 7 above)
Comment (p2004a) [reply docker-build-v2/amd64-linux/Dockerfile:44]: Yes, we already use it, I've switched to it in https://github.com/beyond-all-reason/spring/commit/d26dc0a844e946a3d70a26d46c774215b20febf6 long time ago because it automatically properly handles resolution of static libraries including dependencies of those static libraries.

File: docker-build-v2/build.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (p2004a) [reply docker-build-v2/build.sh]: I've split it and with the split added more functionality to the script: allows to only configure or only compile.

**Review by sprunk — APPROVED** [review]
LGTM in a general way, but I'm not too competent in this area so if you want a proper review probably wait for ivand's

**Comment by p2004a** [comment]
@lhog told me it's good for him, so I will merge
